### PR TITLE
Fix outer reduction performance

### PR DESCRIPTION
**Pull request by naoyam**
Improves performance of outer reductions, particularly for channels last batch norm like normalizations.

### CODE REVIEW

**Review by csarofeen — COMMENTED** [review]
I made a lot of the changes, leaving the review to Naoya.

File: torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp [hunk 1]
```diff
@@ -13,6 +13,8 @@
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp]: This file should be removed from this PR!

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp]: This is from https://github.com/csarofeen/pytorch/pull/1695 and evidently 1695 damages channels first perf.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 2]
```diff
@@ -466,20 +517,6 @@ ReductionParams OuterReductionHeuristic(
     const int64_t n_tensor_inputs,
     const int64_t max_input_dtype_size,
     const size_t vectorize_factor) {
-  // Set some targets for parallelization
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp:469]: Start of outer reduction heuristic changes.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction_heuristic.h [hunk 3]
```diff
@@ -111,7 +115,7 @@ class ReductionParams {
     bool attr_equal = other.fastest_dim == fastest_dim &&
         other.persistent_kernel == persistent_kernel &&
         other.project_persistent_buffers == project_persistent_buffers &&
-        other.schedule_3D == schedule_3D &&
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction_heuristic.h:114]: flip_grid was never used, may want to leave it in to play with later because I didn't test in the end if there are any cases that can benefit. Just please mark with a TODO.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 4]
```diff
@@ -486,9 +486,27 @@ size_t SchedulerRuntimeInfo::getMaxVectorizableWidth(TensorView* tv) {
   // If we don't have an record, either it is a tv with innermost broadcast,
   // or it is an intermediate tensor allocated by fuser. Logic copied to get
   // root according to scheduler_utils::innerMostRootDim.
-  auto tv_root = TensorDomain::noReductions(
-      tv->hasReduction() && tv->hasRFactor() ? tv->getRootDomain()
-                                             : tv->getMaybeRFactorDomain());
+  auto tv_root = tv->hasReduction() && tv->hasRFactor()
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:489]: This was a fun fix.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 5]
```diff
@@ -486,9 +486,27 @@ size_t SchedulerRuntimeInfo::getMaxVectorizableWidth(TensorView* tv) {
...
+      ? tv->getRootDomain()
+      : tv->getMaybeRFactorDomain();
+
+  auto tv_root_no_reductions = TensorDomain::noReductions(tv_root);
+
+  auto contiguity = tv->domain()->contiguity();
+  // Appears after reductions the reduction domain often has a contiguity entry.
+  // This only matters if the result of the reduction is an output
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:497]: A little confused here. Can `contiguity()` sometimes only account for non-reduction domains but other times both reduction and non-reduction domains?

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 6]
```diff
@@ -22698,6 +22700,103 @@ TEST_F(NVFuserMultithreadedTest, MultipleFunctions_CUDA) {
...
+  int TIDx = 16;
+  int vec = 4;
+
+  int TIDy = 16;
+  int outer_tidy_fact = 16;
+
+  reduction_tv->split(-1, TIDx * vec);
+  reduction_tv->split(-1, vec);
+  reduction_tv->axis(-2)->parallelize(ParallelType::TIDx);
+  // TODO: enable:
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp]: Vectorizing `reduction_tv->axis(-1)` can cause a problem? Tried it, nothing seems to have failed.

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:497]: Seems when we reduce the resulting tensor has a contiguity entry for the reduction dims, we should probably fix that.

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/test/test_gpu.cpp]: Oh I think that should be uncommented, that should be fixed now. Can you change it to vectorize and remove unroll?

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp [hunk 7]
```diff
@@ -649,44 +763,47 @@ ReductionParams OuterReductionHeuristic(
...
-    iter_unroll_factor = 1;
-  }
-
   int64_t gdimx = LaunchParams::UNINITIALIZED_VAL;
   int64_t gdimy = LaunchParams::UNINITIALIZED_VAL;
 
+  // In these instances latency of the cleanup may be significant so flip gdimx
+  // and gdimy to try and prevent all cleanup from happening at the same time
+  bool flip_grid = gidim > 1 && gidim < 8;
+  flip_grid = false;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp]: So, currently, are we just always disabling this?

**Comment by naoyam** [comment]
The PR looks good to me, though I have seen vectorization error with `FusionWelfordShmoo`, which I can't repro.

```
[ RUN      ] NVFuserTest.FusionWelfordShmoo_CUDA
unknown file: Failure
C++ exception with description "Tried to vectorize a dim resulting in a word size of 32 however, vector sizes only upto and including 16 bytes are supported.
Exception raised from validate at ../torch/csrc/jit/codegen/cuda/lower_validation.cpp:396 (most recent call first):
frame #0: <unknown function> + 0xa44b0 (0x7f264ea564b0 in /home/nmaruyama/pytorch/debug2/build/lib/libc10.so)
frame #1: std::function<std::__cxx11::basic_string<char, st
```

The error is concerning, but I don't find where the reduction scheduler could set the invalid vectorization size. Although not ideal, this shouldn't cause silent errors but should be detected by the validation. 

Also, `FusionViewPersistentShmoo` would hit invalid memory access errors with compute-sanitizer. Needs #1707.

File: torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp (hunk 7 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp]: Yes, it is piped through correctly to use, but I haven't used it yet as I didn't see any cases benefit. However, I didn't check if any cases benefit after tuning the heuristics, I want to give it a try when I have a chance then will cleanup if it's not useful on outer reductions.

**Review by csarofeen — APPROVED** [review]
LGTM Thanks for the cleanup!

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:497]: Filed an issue #1708
